### PR TITLE
Impossible to define static 'length' function on class

### LINKED ISSUE #442: Impossible to define static 'length' function on class
The following typescript program is accepted:

``` typescript
class C {
    static length ()  { return "twelve"; }
}
```

However, in the generated code:

``` javascript
var C = (function () {
    function C() {
    }
    C.length = function () {
        return "twelve";
    };
    return C;
})();
```

Here, an attempt is made to assign to the `length` property of a function. This doesn't work (at least in Firefox and Chrome), causing subsequent calls of `C.length()` to crash. Perhaps calling a static class function `length` should just be disallowed?

Migrated from codeplex [issue #1260](http://typescript.codeplex.com/workitem/1260).

### PR DESCRIPTION
**Pull request by plantain-00**
Fixes #442

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13175,6 +13177,21 @@ namespace ts {
             if (node.flags & NodeFlags.Abstract && node.body) {
                 error(node, Diagnostics.Method_0_cannot_have_an_implementation_because_it_is_marked_abstract, declarationNameToString(node.name));
             }
+
+            checkNameOfStaticMethodOrPropertyInClass(node);
+        }
+
+        function checkNameOfStaticMethodOrPropertyInClass(node: ClassElement) {
```
Comment (yuit) [inline src/compiler/checker.ts]: I would consider this as grammar check -> like call it `checkGrammarStaticPropertyName`  and move it to the end of the line  [13161](https://github.com/Microsoft/TypeScript/pull/9215/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR13161)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13161,6 +13161,8 @@ namespace ts {
             checkGrammarDecorators(node) || checkGrammarModifiers(node) || checkGrammarProperty(node) || checkGrammarComputedPropertyName(node.name);
 
             checkVariableLikeDeclaration(node);
+
+            checkNameOfStaticMethodOrPropertyInClass(node);
```
Comment (yuit) [inline src/compiler/checker.ts]: I would move it to the end of the line  [13161](https://github.com/Microsoft/TypeScript/pull/9215/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR13161)

**Comment by yuit** [comment]
I just found out something here, I believe there need to be distinction between static property and static method declaration when users target ES5, ES6

- ES5 : **all** (length, name, arguments, caller) should be disallowed for both static property and static method declaration

- ES6: **all** (length, name, arguments, caller) should be disallowed when declared as static property but (length, arguments, caller) should be allow in method declaration ("name" get special treatment [here](http://www.ecma-international.org/ecma-262/6.0/#sec-class-definitions-runtime-semantics-evaluation)). From trying out in node, when user defined static method declaration with those names they seems override the reserved property except "name". I can't find it in the spec :( @bterlson, could you help confirming whether in ES6(length, arguments, caller) are allowed as static method declaration

**Comment by yuit** [comment]
Regardless of the previous comment, @vladima points out that this rule should be extended to "ComputedProperty" with string literal of the value: length, name, arguments, caller

Could you add additional tests for such case?

File: src/compiler/checker.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: We should do this in a more generic fashion and not hard code the names. we should use the `Function` type and make sure we are not redefining any of the properties there. possibly `readonly` ones only. this way users can add/remove some of these checks.

Comment (plantain-00) [inline src/compiler/checker.ts]: @mhegazy and @yuit , thanks for the review, I will find time to fix these.

I'm wondering how to use the `Function` type, `name === nameof Function.length` is not allowed yet, and `Function[name] !== undefined` will not work for the case of `name === "apply"`.

**Comment by plantain-00** [comment]
@yuit `extended to computed property` means errors in below code? 
```ts
class C {
    static caller = { // error
        [name]: "abc", // error
        [length]: 1, // error
    };
}
```

**Comment by yuit** [comment]
@plantain-00  almost but using string literal instead. In your example, it is an expression
```typescript
class C {
    static caller = { // error: caller
        ["name"]: "abc", // error: name
        ["length"]: 1, // error: length
    };
}
```

**Comment by plantain-00** [comment]
any comments?

**Comment by bterlson** [comment]
@yuit sorry I missed my ping. Confirming that this change is correct, length, arguments, caller, and name are all valid static method names in ES6.

File: tests/baselines/reference/checkGrammarStaticMemberName_es5.errors.txt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+!!! error TS2690: 'name' is not allowed to be used as a name of a static property or method in a class.
+        static arguments = 1;
+        ~~~~~~~~~~~~~~~~~~~~~
+!!! error TS2690: 'arguments' is not allowed to be used as a name of a static property or method in a class.
+        static caller = 1;
+        ~~~~~~~~~~~~~~~~~~
+!!! error TS2690: 'caller' is not allowed to be used as a name of a static property or method in a class.
+        static foo = {
+            ["length"]: 1,
+            ~~~~~~~~~~
```
Comment (mhegazy) [inline tests/baselines/reference/checkGrammarStaticMemberName_es5.errors.txt]: why is this an error?

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -1939,6 +1939,10 @@
         "category": "Error",
         "code": 2689
     },
+    "'{0}' is not allowed to be used as a name of a static property or method in a class.": {
```
Comment (mhegazy) [inline src/compiler/diagnosticMessages.json]: this should be split into two error messages, one for properties and methods for target ES5, and one for properties only for ES6 and higher.

File: tests/baselines/reference/checkGrammarStaticMemberName_es5.errors.txt (hunk 3 above)
Comment (plantain-00) [inline tests/baselines/reference/checkGrammarStaticMemberName_es5.errors.txt]: There is a comment(above) from yuit:
> this rule should be extended to "ComputedProperty" with string literal of the value: length, name, arguments, caller

File: src/compiler/diagnosticMessages.json (hunk 4 above)
Comment (plantain-00) [inline src/compiler/diagnosticMessages.json]: Will do, and fix the conflicts meanwhile.

**Comment by plantain-00** [comment]
any comments?

**Comment by mhegazy** [comment]
Fixed by https://github.com/Microsoft/TypeScript/pull/12065

**Pull request closed without merging**